firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +159,91 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
     return Tasks.call(executor, this::deleteFirebaseInstallationId);
   }
 
-  /** Returns the application id of the {@link FirebaseApp} of this {@link FirebaseInstallations} */
-  @VisibleForTesting
-  String getApplicationId() {
-    return firebaseApp.getOptions().getApplicationId();
-  }
-
-  /** Returns the nick name of the {@link FirebaseApp} of this {@link FirebaseInstallations} */
-  @VisibleForTesting
-  String getName() {
-    return firebaseApp.getName();
-  }
-
-  /**
-   * Returns the {@link PersistedFidEntry} from shared prefs.
-   *
-   * @throws {@link FirebaseInstallationsException} when shared pref is empty or {@link
-   *     PersistedFidEntry} is in error state.
-   */
-  private PersistedFidEntry getPersistedFid() throws FirebaseInstallationsException {
-    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
-    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
-      throw new FirebaseInstallationsException(
-          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
+  private Task<String> addGetIdListener() {
+    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+    StateListener l = new GetIdListener(taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
     }
-    return persistedFidEntry;
+    return taskCompletionSource.getTask();
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
+  private Task<InstallationTokenResult> addGetAuthTokenListener(int authTokenOption) {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    String oldAuthToken =
+        persistedFid.readPersistedFidEntryValue().getAuthToken() == null
+            ? ""
+            : persistedFid.readPersistedFidEntryValue().getAuthToken();
```

Comment:
this doesn't even have to be ints, it could probably just be a boolean ("shouldRefresh"?) that's accessed in the same synchronized block as the listeners additions

Reply:
About the file I/O : shouldn't it be okay even if its called by any thread? As long as our onStateReached conditional can handle it correctly. I tested it with my test cases and it seemed to work fine. However, you are right about https://github.com/firebase/firebase-android-sdk/pull/847#discussion_r329233840 .

I tried implementing your suggestion based on my understanding of the comment. Also, I have tested the changes. I will be honest, I dint understand this part 

> ("shouldRefresh"?) that's accessed in the same synchronized block as the listeners additions

please let me know wdyt.